core/src/main/java/hudson/FilePath.java

```diff
@@ -1105,23 +1104,28 @@ protected void after() {}
      * @since 1.522
      */
     public <V> Callable<V,IOException> asCallableWith(final FileCallable<V> task) {
-        return new Callable<V,IOException>() {
-            @Override
-            public V call() throws IOException {
-                try {
-                    return act(task);
-                } catch (InterruptedException e) {
-                    throw (IOException)new InterruptedIOException().initCause(e);
-                }
+        return new CallableWith<>(task);
+    }
+    private class CallableWith<V> implements Callable<V, IOException> {
```

Comment:
You may be surprised to see a number of these nested classes being non`static`. This was just the simplest refactoring that addresses the essential problems of using anonymous inner classes as described in https://github.com/jenkinsci/jenkins/pull/3312#issuecomment-368713851. You could make the nested classes `static`, but then you would need to define a `FilePath` field anyway, which just makes the code more verbose and does nothing beyond rename the implicit `this$0` field.

So why do all these nested classes inside `FilePath` _need_ that field to be serialized, you might wonder? Well, for its `SoloFilePathFilter filter` field, passed to calls like `reading`, delegating to methods like `FilePathFilter.read`.

OK, so why not define a `SoloFilePathFilter filter` field in a `static` nested class then? Well, because it is not `Serializable`.

Huh? How can it not be serializable, if it is a field we are transferring to the remote side and then using? In fact it is `transient`. And it is restored in `FilePath.readObject` by some complicated logic which takes into account the serial form of `FilePath`.

Basically, this would only be possible if `FilePathFilter` were given a `writeReplace` method which produced some sort of placeholder object based on `Channel.current() == null` which `readResolve`d to either `null` or `SoloFilePathFilter.wrap(FilePathFilter.current())`. A bigger refactoring than I felt comfortable with.